go/indexer/db/db.go

```diff
@@ -0,0 +1,657 @@
+		}
+
+		if rows.Next() {
+			return errors.New("number of rows should be at most 1 since LIMIT is 1")
+		}
+
+		highestBlock = &L2BlockLocator{
+			Number: number,
+			Hash:   l2common.HexToHash(hash),
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return highestBlock, nil
+}
+
+func (d *Database) GetIndexedL1BlockByHash(hash common.Hash) (*IndexedL1Block, error) {
+	const selectBlockByHashStatement = `
+	SELECT
+		hash, parent_hash, number, timestamp
+	FROM l1_blocks
+	WHERE hash = $1
+	`
+
+	var block *IndexedL1Block
+	err := txn(d.db, func(tx *sql.Tx) error {
+		queryStmt, err := tx.Prepare(selectBlockByHashStatement)
+		if err != nil {
+			return err
+		}
+
+		rows, err := queryStmt.Query(hash.String())
+		if err != nil {
+			return err
+		}
+
+		if !rows.Next() {
+			return nil
+		}
+
+		var hash string
+		var parentHash string
+		var number uint64
+		var timestamp uint64
+		err = rows.Scan(&hash, &parentHash, &number, &timestamp)
+		if err != nil {
+			return err
+		}
+
+		block = &IndexedL1Block{
+			Hash:       common.HexToHash(hash),
+			ParentHash: common.HexToHash(parentHash),
+			Number:     number,
+			Timestamp:  timestamp,
+			Deposits:   nil,
+		}
+
+		if rows.Next() {
+			return errors.New("number of rows should be at most 1 since hash is pk")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return block, nil
+
+}
+func (d *Database) GetEventsByBlockHash(hash common.Hash) ([]TxnEnqueuedEvent, error) {
+	const selectEventsByBlockHashStatement = `
+	SELECT
+		b.number, b.timestamp,
+		d.tx_hash, d.data
+	FROM
+		blocks AS b,
+		deposits AS d
+	WHERE b.hash = d.block_hash AND b.hash = $1
+	`
+
+	var events []TxnEnqueuedEvent
+	err := txn(d.db, func(tx *sql.Tx) error {
+		queryStmt, err := tx.Prepare(selectEventsByBlockHashStatement)
+		if err != nil {
+			return err
+		}
+
+		rows, err := queryStmt.Query(hash.String())
+		if err != nil {
+			return err
+		}
+
+		for rows.Next() {
+			event, err := scanTxnEnqueuedEvent(rows)
+			if err != nil {
+				return err
+			}
+
+			events = append(events, event)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return events, nil
+}
+
+func scanTxnEnqueuedEvent(rows *sql.Rows) (TxnEnqueuedEvent, error) {
+	var number uint64
+	var timestamp uint64
+	var txHash string
+	var data []byte
+	err := rows.Scan(
+		&number,
+		&timestamp,
+		&txHash,
+		&data,
+	)
+	if err != nil {
+		return TxnEnqueuedEvent{}, err
+	}
+
+	return TxnEnqueuedEvent{
+		BlockNumber: number,
+		Timestamp:   timestamp,
+		TxHash:      common.HexToHash(txHash),
+		Data:        data,
+	}, nil
+}
+
+func (d *Database) Close() error {
+	return d.db.Close()
+}
+
+func (d *Database) Config() string {
+	return d.config
+}
+
+func NewDatabase(config string) (*Database, error) {
+	db, err := sql.Open("postgres", config)
+	if err != nil {
+		return nil, err
+	}
+
+	err = db.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, migration := range schema {
```

Comment:
IMO, we should have a special `migrate` subcommand rather than doing this on every startup. Otherwise, load balanced instances of the indexer could all try and migrate at once. You'd do:

`indexer migrate --db-host ...`